Src/LiquidProjections.PollingEventStore/PollingEventStoreAdapter.cs

```diff
@@ -134,276 +210,190 @@ private Page TryGetNextPageFromCache(long precedingCheckpoint, string subscripti
-                        from requesting already loaded transactions which are not added to cache yet. */
-                    for (int index = transactions.Count - 1; index > 0; index--)
-                    {
-                        cachedTransactionsByPrecedingCheckpoint.Set(transactions[index - 1].Checkpoint, transactions[index]);
-                    }
-
-                    cachedTransactionsByPrecedingCheckpoint.Set(precedingCheckpoint, transactions[0]);
-
-#if LIQUIDPROJECTIONS_DIAGNOSTICS
-                    logger(() =>
-                        $"Loader for subscription {subscriptionId ?? "without ID"} has cached {transactions.Count} transactions " +
-                        $"from checkpoint {transactions.First().Checkpoint} to checkpoint {transactions.Last().Checkpoint}.");
-#endif
-                }
             }
             else
             {
-#if LIQUIDPROJECTIONS_DIAGNOSTICS
-                logger(() =>
-                    $"Loader for subscription {subscriptionId} has discovered " +
-                    $"that there are no new transactions yet. Next request for the new transactions will be delayed.");
-#endif
+                logDebug(() =>
+                    $"No Transactions: (subscriptionId: {subscriptionId}, precedingCheckpoint: {precedingCheckpoint})");
             }
 
             return new Page(precedingCheckpoint, transactions);
         }
 
+        private void CachePage(Page page, string subscriptionId)
+        {
+            IReadOnlyList<Transaction> transactions = page.Transactions;
+
+            if (cachedTransactionsByPrecedingCheckpoint != null)
+            {
+                /* Add to cache in reverse order to prevent other projectors
+                    from requesting already loaded transactions which are not added to cache yet. */
+                for (int index = transactions.Count - 1; index > 0; index--)
+                {
+                    cachedTransactionsByPrecedingCheckpoint.Set(transactions[index - 1].Checkpoint, transactions[index]);
+                }
+
+                cachedTransactionsByPrecedingCheckpoint.Set(page.PrecedingCheckpoint, transactions[0]);
+
+                logDebug(() =>
+                    $"Cached: (subscription: {subscriptionId}, precedingCheckpoint: {page.PrecedingCheckpoint}, transactions: {transactions.Count}, range: {transactions.First().Checkpoint}-{transactions.Last().Checkpoint}.");
+            }
+        }
+
         public void Dispose()
         {
-            lock (SubscriptionLock)
+            lock (subscriptionLock)
             {
-                if (!isDisposed)
+                if (!disposingCancellationTokenSource.IsCancellationRequested)
                 {
-                    isDisposed = true;
+                    disposingCancellationTokenSource.Cancel();
 
-                    foreach (Subscription subscription in Subscriptions.ToArray())
+                    foreach (Subscription subscription in subscriptions.ToArray())

```

Comment:
🔧 Since you already lock the subscriptions hashset when modifying it, you probably don't need the ToArray. (Or could you maybe remove the lock?)